README.md
# README
## MeteorHelper - Meteor CLI integration into VSCode

Use this VSCode Extension to integrate the Meteor CLI from within Visual Studio Code.

## Installation

1. Press `F1` to bring up Command Palette
2. Search for `Extensions: Install Extension` and select the command
3. Type in 'MeteorHelper' and select the extension to install. 
4. Restart VSCode to activate MeteorHelper.

## Usage

* Optionally specify the MeteorHelper configuration in your workspace settings.json file (all properties begin with the 'meteorhelper' prefix
* Use the MeteorHelper commandbar to start/stop Meteor and to start testing.
* Use the MeteorHelper commands (available via the `F1` key) to perform several Meteor actions like running and stopping Meteor, performing resets etc. 

## List of Meteor commands supported

The current version of MeteorHelper supports the following meteor commands:

* Meteor add
* Meteor debug
* Meteor list
* Meteor npm
* Meteor remove
* Meteor reset
* Meteor run
* Meteor test
* Meteor update

## Relevant Workspace Configuration

The configuration properties stated below can be added to the VSCode configuration to configure settings for Meteor.

```
* meteorhelper.meteorPath: path to the Meteor executable (e.g. /usr/local/bin/meteor on the Mac or %USERPROFILE%\\AppData\\Local\\.meteor\\meteor.bat on Windows
* meteorhelper.meteorPort: port to be used when running the Meteor process (defaults to 3000)
* meteorhelper.meteorSettings: Relative path to the Meteor settings JSON file, e.g. app/settings/settings.debug.json
* meteorhelper.relativeProjectPath: Relative path to your Meteor application, e.g. app
* meteorhelper.envArgs: environment variables to be used with Meteor
* meteorhelper.testConfiguration: test configuration settings
```

 The environment variables Meteor supports, are listed on the [environment-variables.html](https://docs.meteor.com/environment-variables.html) page. Example usage with the envArgs configuration property:
 
```
 "meteorhelper.envArgs": [
        { "argName": "MAIL_URL", "argValue": "smtp://user:pass@yourservice.com:587" },
        { "argName": "MONGO_URL", "argValue": "mongodb://user:password@myserver.com:10139" }
    ]
```

### Meteor Test

To enable testing (meteor test), Meteor 1.3 or above is required. Also, there needs to be a test driver package added into the project. Next, specify the driver package to be used in the driverPackage property under the meteorhelper.testConfiguration workspace section.

In MeteorHelper 0.1.4 the ability is added to specify additional configuration for testing via environment variables. You can use the MeteorHelper workspace configuration to specify these key/value pairs by adding them to the envArgs property under the meteorhelper.testConfiguration section. 

For example, to add 'TEST_WATCH=1' to 'meteor test' (i.e. you want to execute ````TEST_WATCH=1 meteor test --driver-package meteortesting:mocha'````) as mentioned in the [meteor-mocha](https://github.com/meteortesting/meteor-mocha) package, you use the envArgs property like this: 

````
"meteorhelper.testConfiguration": {
    "driverPackage": "meteortesting:mocha",
    "debugPort": "3001",
    "envArgs": [
        {
            "argName": "TEST_WATCH", "argValue": "1"
        }
    ]
}
````

### For more information
* [Meteor Docs](http://docs.meteor.com/#/full/meteorhelp)
* [Meteor](https://www.meteor.com)
* See [Meteor Guide](http://guide.meteor.com/testing.html) for more information about testing.

## Contact Information ##

* [https://twitter.com/ramonitor](https://twitter.com/ramonitor)





CHANGELOG.md
# CHANGELOG
## MeteorHelper - Meteor CLI integration into VSCode

## 0.1.5
* Added the envArgs configuration options for Meteor Run and Debug (as an alternative and more generic solution to specify a MONGO_URL as mentioned in PR [#8](https://github.com/ramonitor/vscode-meteorhelper/pull/8), thanks [@ntziolis](https://github.com/ntziolis))
* Removed dashed lines from MeteorHelper output to prevent issues with smaller VSCode output windows (Issue [#10](https://github.com/ramonitor/vscode-meteorhelper/issues/10), thanks [@laurentenhoor](https://github.com/laurentenhoor))

## 0.1.4
* Added the ability to specify additional testing configuration via the envArgs property of the workspace configuration (Issue [#6](https://github.com/ramonitor/vscode-meteorhelper/issues/6) and Issue [#7](https://github.com/ramonitor/vscode-meteorhelper/issues/7))
* Added support for the 'meteor test' --extra-packages parameter 

## 0.1.3
* Fixed package.json structure of the extension causing it to fail (Issue [#2](https://github.com/ramonitor/vscode-meteorhelper/issues/2))

### 0.1.2
* Running Meteor processes are killed when VSCode is being shutdown (Issue [#1](https://github.com/ramonitor/vscode-meteorhelper/issues/1))

### 0.1.1

* Fixed a bug with the meteor reset command
* Added validation of the Meteor project directory 
* MeteorHelper extension can be activated via the supported MeteorHelper commands

### 0.1.0

* Added support for the Meteor commands add, remove, list, update, npm and test
* Added support for Meteor testing (requires Meteor 1.3 or higher)
* Added multiple Output Channels (meteor run, test and other commands run in separate output channels)
* Added Meteor version detection
* Several optimizations

### 0.0.3

* Project cleanup
* Minor code optimization
* Code on Github

### 0.0.2

* Meteor URL fix

### 0.0.1

* Initial version

** Enjoy!**


src/extension.ts
import * as vscode from 'vscode';
import { MeteorCommandHelper } from './mh/MeteorCommandHelper';
import { ConfigHelper } from './mh/ConfigHelper';

export function activate(context: vscode.ExtensionContext) {
    console.log('MeteorHelper activated..');

    MeteorCommandHelper.init();

    context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(() => {
        ConfigHelper.setConfiguration();
    }));

    context.subscriptions.push(MeteorCommandHelper.registerRunCommand('meteorhelper.meteorRun', 'run'));
    context.subscriptions.push(MeteorCommandHelper.registerRunCommand('meteorhelper.meteorDebug', 'debug'));

    context.subscriptions.push(MeteorCommandHelper.registerTestCommand('meteorhelper.meteorTest', 'test'));
    context.subscriptions.push(MeteorCommandHelper.registerTestCommand('meteorhelper.meteorTestFullApp', 'test', 'full-app'));

    context.subscriptions.push(MeteorCommandHelper.stopCommand('meteorhelper.meteorStop'));
    context.subscriptions.push(MeteorCommandHelper.stopRunTestCommand('meteorhelper.meteorStopRun', 'run'));
    context.subscriptions.push(MeteorCommandHelper.stopRunTestCommand('meteorhelper.meteorStopTest', 'test'));

    context.subscriptions.push(MeteorCommandHelper.registerCommand('meteorhelper.meteorReset', 'reset'));
    context.subscriptions.push(MeteorCommandHelper.registerCommand('meteorhelper.meteorList', 'list'));

    context.subscriptions.push(MeteorCommandHelper.registerCommandWithUserInput('meteorhelper.meteorAdd', false, 'add'));
    context.subscriptions.push(MeteorCommandHelper.registerCommandWithUserInput('meteorhelper.meteorRemove', false, 'remove'));
    context.subscriptions.push(MeteorCommandHelper.registerCommandWithUserInput('meteorhelper.meteorUpdate', true, 'update'));
    context.subscriptions.push(MeteorCommandHelper.registerCommandWithUserInput('meteorhelper.meteorNpm', false, 'npm'));
}

export function deactivate() {
    const runningCommands = MeteorCommandHelper.getCommandsFromExecutionList();
    
    runningCommands.forEach(command => {
        console.log('MeteorHelper: killing PID ', command.meteorCommand.pid);
        command.meteorCommand.kill();
    });

    console.log('MeteorHelper deactivated!');
}




src/mh/ConfigHelper.ts
import * as vscode from 'vscode';
const path = require('path');
const fs = require('fs');
const shell = require('shelljs');

interface ITestConfiguration {
	debugPort: string;
	driverPackage: string;
	extraPackages: Array<string>;
	port: string;
	verbose: boolean;
	settings: string;
	envArgs: [IKeyValueConfigItem];
}

interface IKeyValueConfigItem {
	argName: string;
	argValue: string;
}

export class ConfigHelper {
	private static CONFIG_SECTION = 'meteorhelper';
	private static EXTENSION_NAME = 'ramonitor.meteorhelper';
	private static config: vscode.WorkspaceConfiguration;

	private static commandConfig = {
		commands: [
			{
				command: 'meteorhelper.meteorRun',
				prompt: 'meteor run',
				placeHolder: '[target..] [options]'
			},
			{
				command: 'meteorhelper.meteorAdd',
				prompt: 'meteor add',
				placeHolder: '<package> [package..]'
			},
			{
				command: 'meteorhelper.meteorList',
				prompt: 'meteor list',
				placeHolder: '<package> [package..]'
			},
			{
				command: 'meteorhelper.meteorUpdate',
				prompt: 'meteor update',
				placeHolder: 'specify list of packagenames or press enter for an update of the meteor release'
			},
			{
				command: 'meteorhelper.meteorRemove',
				prompt: 'meteor remove',
				placeHolder: '[packageName packageName2 ...]'
			},
			{
				command: 'meteorhelper.meteorNpm',
				prompt: 'meteor npm',
				placeHolder: '<command>'
			}
		]
	};

	public static setConfiguration(): void {
		this.config = vscode.workspace.getConfiguration(this.CONFIG_SECTION);
	}

	public static getCommandConfig(command: string): any {
		return this.commandConfig.commands.find((entry) => entry.command == command);
	}

	public static getMeteorPath(): string {
		const defaultPaths = [
			{
				platform: 'win32',
				path: '%USERPROFILE%\\AppData\\Local\\.meteor\\meteor.bat'
			},
			{
				platform: 'darwin',
				path: '/usr/local/bin/meteor'
			}
		];

		let meteorPath = this.config.get('meteorPath', '');

		if (!meteorPath) {
			const installPath = defaultPaths.find((obj) => obj.platform === process.platform).path;

			if (process.platform === 'win32') {
				meteorPath = installPath.replace(/%([^%]+)%/g, function(_, n) {
					return process.env[n];
				});
			} else {
				meteorPath = installPath;
			}
		}

		return meteorPath;
	}

	public static getMeteorSettingsFile(): string {
		return this.config.get('meteorSettings', '');
	}

	public static getMeteorPort(): string {
		return this.config.get('meteorPort', '');
	}

	public static getEnvArgsConfig(): IKeyValueConfigItem[] {
		return this.config.get('envArgs');
	}

	public static getMeteorAppPath(): string {
		const projectPath = vscode.workspace.rootPath;
		const relativeProjectPath = this.config.get('relativeProjectPath', '');

		const meteorAppPath = relativeProjectPath ? path.join(projectPath, relativeProjectPath) : projectPath;

		return meteorAppPath;
	}

	public static isMeteorProjectFolder(projectFolder: string): boolean {
		return shell.test('-e', path.join(projectFolder, '.meteor'));
	}

	public static getExtensionVersion(): string {
		const extension = vscode.extensions.getExtension(this.EXTENSION_NAME);

		if (extension) {
			return extension.packageJSON.version;
		} else {
			return 'unknown';
		}
	}

	public static getMeteorTestConfig(): ITestConfiguration {
		const testConfig = this.config.get<ITestConfiguration>('testConfiguration');
		return testConfig;
	}

	public static getMeteorMajorVersion(): number {
		const release = '.meteor/release';
		const releasefile = path.join(ConfigHelper.getMeteorAppPath(), release);

		if (!shell.test('-f', releasefile)) {
			throw new Error('MeteorHelper: Meteor release file not found - version could not be determined.');
		}

		const releaseVersion: string = fs.readFileSync(releasefile, 'utf8').split('@')[1].substr(0, 3);
		return parseFloat(releaseVersion);
	}

	public static mapKeyValueConfigToEnvArgs(keyValueItems: IKeyValueConfigItem[]): Object {
		const envArgsObject = keyValueItems.reduce(
			(obj, item) => Object.assign(obj, { [item.argName.toUpperCase()]: item.argValue }),
			{}
		);

		return envArgsObject;
	}

	public static mapEnvArgsToString(keyValueObject: Object): string {
		const envArgsString = Object.keys(keyValueObject).map((kv) => [ kv, keyValueObject[kv] ].join('=')).join(' ');
		return envArgsString;
	}
}


src/mh/StatusBarHelper.ts
import * as vscode from 'vscode';

export enum RunMode {
    Running,
    Debugging,
    Stopped
}

export enum TestMode {
    Normal,
    Full,
    Stopped
}

export class StatusBarHelper implements vscode.Disposable {
    private meteorRunDebugStatusBar: vscode.StatusBarItem;
    private meteorTestStatusBar: vscode.StatusBarItem;
    private statusTextBar: vscode.StatusBarItem;
    private statusTextBarTimeout = 5000;

    constructor() {
        this.statusTextBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 3);
        this.meteorRunDebugStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 2);
        this.meteorTestStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1);
    }

    public dispose(): void {
        this.meteorRunDebugStatusBar.dispose();
        this.meteorTestStatusBar.dispose();
        this.statusTextBar.dispose();
    }
    
    public init(initialMessage): void {
        this.setMessage(initialMessage, true);
        
        this.setMeteorRunStatusBar(RunMode.Stopped, 'Click to start meteor run', 'meteorhelper.meteorRun');
        this.setMeteorTestStatusBar(TestMode.Stopped, 'Click to start meteor test', 'meteorhelper.meteorTest');

        this.meteorRunDebugStatusBar.show();
        this.meteorTestStatusBar.show();
    }

    setMeteorRunStatusBar(state: RunMode, tooltip?: string, command?: string): void {
        tooltip = tooltip || '';
        let text;

        switch (state) {
            case RunMode.Running:
                text = `Meteor: Running`;
                break;
            case RunMode.Debugging:
                text = `Meteor: Debugging`;
                break;
            case RunMode.Stopped:
                text = `Meteor: Idle`;
                break;
        }
        
        this.meteorRunDebugStatusBar.text = text;
        this.meteorRunDebugStatusBar.tooltip = tooltip;
        this.meteorRunDebugStatusBar.command = command;
    }

    setMeteorTestStatusBar(state: TestMode, tooltip?: string, command?: string): void {
        tooltip = tooltip || '';
        let text;

        switch (state) {
            case TestMode.Normal:
                text = `Meteor Test: Running`;
                break;
            case TestMode.Full:
                text = `Meteor Test: Running Full`;
                break;
            case TestMode.Stopped:
                text = `Meteor Test: Idle`;
                break;
        }

        this.meteorTestStatusBar.text = text;
        this.meteorTestStatusBar.tooltip = tooltip;
        this.meteorTestStatusBar.command = command;
    }

    setMessage(message: string, hideAfterTimeout?: boolean, color?: string): Thenable<boolean> {
        return new Promise((resolve, reject) => {
            this.statusTextBar.text = message;
            this.statusTextBar.color = color || null;
            this.statusTextBar.show();

            if (hideAfterTimeout) {
                setTimeout(() => {
                    if (this.statusTextBar.text == message) {
                        this.statusTextBar.hide();
                    }
                    resolve(true);
                }, this.statusTextBarTimeout);
            } else {
                resolve(true);
            }
        });
    }

}

src/meteorhelper.ts
'use strict';

import * as vscode from 'vscode';
import * as meteor from './meteorhelper';
import { exec, spawn, ChildProcess } from 'child_process';
var shell = require('shelljs');
var kill = require('tree-kill');
var path = require('path');

interface IMeteorCommandArgument {
    command: string,
    optionArguments?: IMeteorOptionArgument[]
} 

interface IMeteorOptionArgument {
    option: string,
    value: string
}

enum QuickCommandBarState {
    Stopped = 0,
    Running = 1
}

export enum MeteorHelperCommand {
    Run, 
    Stop,
    Reset,
    Debug,
    Action
}

export class MeteorHelper {

    private static _instance: MeteorHelper = new MeteorHelper();
    private _extension = vscode.extensions.getExtension('ramonitor.meteorhelper');
    private _quickCommandBar: vscode.StatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1);
    private _quickCommandBarState: QuickCommandBarState = QuickCommandBarState.Stopped;
    private _outputChannel: vscode.OutputChannel = vscode.window.createOutputChannel('meteorhelper');
    private _config: vscode.WorkspaceConfiguration;
    private _childProcess: ChildProcess;
    private _configSection: string = 'meteorhelper';
    private _quickCommandsObject: { [state: number]: { 'command': string, 'caption': string, 'tooltip'?: string }; } = {};
    private _messageTimeout: number = 5000;
    private _disposable: vscode.Disposable;    

    constructor() {
        if (MeteorHelper._instance) {
            throw new Error("error");
        }
        MeteorHelper._instance = this;

        this._quickCommandsObject[QuickCommandBarState.Stopped] = { command: 'extension.meteorRun', caption: 'Run Meteor', tooltip: 'MeteorHelper: Run Meteor' };
        this._quickCommandsObject[QuickCommandBarState.Running] = { command: 'extension.meteorStop', caption: 'Stop Meteor', tooltip: 'MeteorHelper: Stop Meteor' };

        this.setMeteorHelperConfig(this._configSection);
        this.checkWorkspaceConfigurationAndShowUser();

        this._disposable = vscode.window.setStatusBarMessage(`MeteorHelper: ${this._extension.packageJSON.version} loaded`, this._messageTimeout);
        this.setQuickCommandBar(QuickCommandBarState.Stopped);
    }

    public static getInstance(): MeteorHelper {
        return MeteorHelper._instance;
    }

    public invokeMeteorCommand(command: MeteorHelperCommand) {

        this.setMeteorHelperConfig(this._configSection);
        this.checkWorkspaceConfigurationAndShowUser();
        
        switch (command) {
            case MeteorHelperCommand.Run:
                this.initMeteorRunOrDebug('run');
                break;
            case MeteorHelperCommand.Stop:
                this.initMeteorStop();
                break;
            case MeteorHelperCommand.Reset:
                this.initMeteorReset();
                break;
            case MeteorHelperCommand.Debug:
                this.initMeteorRunOrDebug('debug');
                break;
            default:
                console.log('not implemented');
                break;
        }
    }

    private initMeteorRunOrDebug(runType: string): void {
        let meteorPath = this.getMeteorPath();
        let meteorSettings = this.getMeteorSettingsFile();
        let meteorPort = this.getMeteorPort();
        let projectPath = this.getMeteorAppPath();

        let meteorRunArgument: IMeteorCommandArgument = { command: runType };
        let meteorOptionsArguments: IMeteorOptionArgument[] = [];

        if (meteorSettings) meteorOptionsArguments.push({ option: '--settings', value: meteorSettings });
        if (meteorPort) meteorOptionsArguments.push({ option: '--port', value: meteorPort });

        meteorRunArgument.optionArguments = meteorOptionsArguments

        if (this.isMeteorRunning()) {
            this._disposable = vscode.window.setStatusBarMessage('MeteorHelper: Meteor already running, stop Meteor first..', this._messageTimeout);
            return;
        }
        
        this.execMeteorProcess(meteorPath, projectPath, meteorRunArgument).then(
            (successMessage: string) => {
                this.setQuickCommandBar(QuickCommandBarState.Running);
                this._disposable = vscode.window.setStatusBarMessage(`MeteorHelper: Running (PID: ${this._childProcess.pid})`);
            },
            (errorMessage: string) => {
                this.setQuickCommandBar(QuickCommandBarState.Stopped);
                vscode.window.setStatusBarMessage('MeteorHelper: Error $(alert): ' + errorMessage);
            });
    }
    
    private initMeteorReset(): void {
        let meteorPath = this.getMeteorPath();
        let projectPath = this.getMeteorAppPath();
        let meteorArgument: IMeteorCommandArgument = { command: 'reset' };
        
        this.askUserConfirmation('Reset Meteor Project State', 'Abort Reset', 'Reset Meteor Project State')
            .then((confirmed: boolean) => {
                    this.initMeteorStop()
                        .then(() => this.execMeteorProcess(meteorPath, projectPath, meteorArgument)
                        .then(() => this._disposable = vscode.window.setStatusBarMessage('MeteorHelper: Meteor reset..', this._messageTimeout)));
        }, () => {
            this._disposable = vscode.window.setStatusBarMessage('MeteorHelper: Meteor reset aborted..', this._messageTimeout);
        });
    }

    private initMeteorStop(): Thenable<boolean> {
        this._disposable.dispose();
        
        return new Promise((resolve, reject) => {
            if (this.isMeteorRunning()) {
                this.killProcessTree(this._childProcess.pid)
                    .then((success: string) => {
                        vscode.window.setStatusBarMessage('MeteorHelper: Meteor stopped', this._messageTimeout);
                        resolve(true);
                    },((errorMessage: string) => {
                        this._disposable = vscode.window.setStatusBarMessage('MeteorHelper: Killing Meteor process failed : ' + errorMessage, this._messageTimeout);
                        reject(false);
                    }) 
            )} else {
                console.log('not running');
                resolve(false);                   
            }
        })
    }

    private execMeteorProcess(meteorPath: string, projectPath: string, meteorArgument: IMeteorCommandArgument): Thenable<string> {
        return new Promise((resolve, reject) => {
            this._outputChannel.clear();
            let parameters: string[] = this.getCommandLineParameters(meteorArgument);
            
            try {
                this._childProcess = spawn(meteorPath, parameters, { cwd: projectPath });
                this._outputChannel.show(vscode.ViewColumn.Three);
                
                if (this.isMeteorRunning()){
                    this.setProcessOutputChannel();
                    resolve('Started Meteor with PID : ' + this._childProcess.pid.toString());   
                } 
            } catch (e) {
                reject (e);
            }
        });
    }
    
    private askUserConfirmation(matchedValue: string, ...options: string[]): Thenable<boolean> {
        return new Promise((resolve, reject) => {
            vscode.window.showQuickPick(options).then((value) => {
                if (value === matchedValue) { 
                    resolve(true);
                } else { 
                    reject(false); 
                }
            });
        });
    }

    private getCommandLineParameters(meteorArgument: IMeteorCommandArgument): string[] {
        let parameters: string[] = [];
        parameters.push(meteorArgument.command);

        if (meteorArgument.optionArguments != undefined) {
            meteorArgument.optionArguments.forEach((item) => {
                parameters.push(item.option)
                parameters.push(item.value);
            });
        }

        return parameters;
    }

    private setProcessOutputChannel(): void {
        this._childProcess.stdout.on('data', (data: Buffer) => {
            if (data) {
                this._outputChannel.append(data.toString());
            }
        });

        this._childProcess.stderr.on('data', (data: Buffer) => {
            if (data) {
                this._outputChannel.append(data.toString());
            }
        });

        this._childProcess.on('exit', (exitcode) => {
            this._outputChannel.append('MeteorHelper: Process exited');
            this._disposable = vscode.window.setStatusBarMessage('MeteorHelper: Ready', this._messageTimeout);
            this.setQuickCommandBar(QuickCommandBarState.Stopped);
            this._childProcess = null;
        });
    }

    private isMeteorRunning(): boolean {
        let isRunning = this._childProcess != undefined;
        return isRunning;
    }

    private killProcessTree(pid: number): Thenable<string> {
        return new Promise((resolve, reject) => {
            try {
                kill(pid, 'SIGTERM', (err) => {
                    if (!err) {
                        this._childProcess = null;
                        resolve('Killed PID ' + pid);
                    }
                    else {
                        reject(err);
                    }
                });
            } catch (e) {
                reject(e);
            }
        })
    } 

    private setQuickCommandBar(state?: QuickCommandBarState): void {
        this._quickCommandBarState = state;
        this._quickCommandBar.command = this._quickCommandsObject[this._quickCommandBarState].command;
        this._quickCommandBar.text = this._quickCommandsObject[this._quickCommandBarState].caption;
        this._quickCommandBar.tooltip = this._quickCommandsObject[this._quickCommandBarState].tooltip;
        this._quickCommandBar.show();
    }

    private getMeteorPort(): string {
        let meteorPort = this._config.get('meteorPort', '3000');
        return meteorPort;
    }

    private getMeteorPath(): string {
        let defaultPaths = [{ 
                'platform' : 'win32', 
                'path': '%USERPROFILE%\\AppData\\Local\\.meteor\\meteor.bat'
            }, { 
                'platform' : 'darwin', 
                'path': '/usr/local/bin/meteor'
            }];
        
        let meteorPath = this._config.get('meteorPath', '');
        
        if (!meteorPath) {
            let installPath = defaultPaths.find((obj) => obj.platform === process.platform).path;

            if (process.platform === 'win32') {
                meteorPath = installPath.replace(/%([^%]+)%/g, function(_,n) {
                    return process.env[n];
                });
            } else {
                meteorPath = installPath;
            }
        }
        
        return meteorPath;
    }

    private getMeteorSettingsFile(): string {
        let meteorSettings = this._config.get('meteorSettings', '');
        return meteorSettings;
    }
    
    private setMeteorHelperConfig(section: string): void {
        this._config = vscode.workspace.getConfiguration(section);
        if (!this._config) {
            throw new Error("MeteorHelper: Error getting configuration");
        }
    }
    
    private getMeteorAppPath() : string {
        let relativeProjectPath = this._config.get('relativeProjectPath', '');
        let meteorAppPath;
        let projectPath = vscode.workspace.rootPath;
        
        if (relativeProjectPath) {
            meteorAppPath = path.join(projectPath, relativeProjectPath);
        } else {
            meteorAppPath = projectPath;
        }
        
        return meteorAppPath;
    }
    
    private checkWorkspaceConfigurationAndShowUser() : void {
        let errorMessages: string [] = this.getWorkspaceConfigurationErrors();
        
        if (errorMessages.length > 0) {
            vscode.window.setStatusBarMessage('MeteorHelper: configuration error $(x)');
            vscode.window.showErrorMessage(errorMessages.join(" "), 'Open Workspace Settings')
                .then(() => vscode.commands.executeCommand('workbench.action.openWorkspaceSettings'));
            
            throw new Error ("MeteorHelper: Workspace Configuration Errors");
        }
    }

    private getWorkspaceConfigurationErrors(): string[] {
        let errorMessages: string[] = [];

        let meteorPath = this.getMeteorPath();
        let meteorSettings = this.getMeteorSettingsFile();
        let projectPath = this.getMeteorAppPath();

        if (!shell.test('-f', meteorPath)) {
            errorMessages.push(`Path to Meteor executable not found`);
        }

        let meteorSettingsFullPath = path.join(projectPath, meteorSettings);
        if (meteorSettings && !shell.test('-f', meteorSettingsFullPath)) {
            errorMessages.push(`Path to settings file incorrect`);
        }

        if (errorMessages.length > 0) errorMessages.unshift('MeteorHelper configuration errors:');

        return errorMessages;
    }
    
}
